src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenRefReadonlyParameterTests.cs

```diff
@@ -0,0 +1,558 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen;
+
+public class CodeGenRefReadonlyParameterTests : CSharpTestBase
+{
+    private const string RequiresLocationAttributeName = "RequiresLocationAttribute";
+    private const string RequiresLocationAttributeNamespace = "System.Runtime.CompilerServices";
+    private const string RequiresLocationAttributeQualifiedName = $"{RequiresLocationAttributeNamespace}.{RequiresLocationAttributeName}";
+
+    private static void VerifyRequiresLocationAttributeSynthesized(ModuleSymbol module)
+    {
+        var attributeType = module.GlobalNamespace.GetMember<NamedTypeSymbol>(RequiresLocationAttributeQualifiedName);
+        if (module is SourceModuleSymbol)
+        {
+            Assert.Null(attributeType);
+        }
+        else
+        {
+            Assert.NotNull(attributeType);
+        }
+    }
+
+    [Fact]
+    public void Method()
+    {
+        var source = """
+            class C
+            {
+                public void M(ref readonly int p) { }
+            }
+            """;
+        var verifier = CompileAndVerify(source, targetFramework: TargetFramework.NetStandard20,
+            sourceSymbolValidator: verify, symbolValidator: verify);
+        verifier.VerifyDiagnostics();
+        verifier.VerifyMethodIL("C", "M", """
+            .method public hidebysig 
+            	instance void M (
+            		[in] int32& p
+            	) cil managed 
+            {
+            	.param [1]
+            		.custom instance void System.Runtime.CompilerServices.RequiresLocationAttribute::.ctor() = (
+            			01 00 00 00
+            		)
+            	// Method begins at RVA 0x2067
+            	// Code size 1 (0x1)
+            	.maxstack 8
+            	IL_0000: ret
+            } // end of method C::M
+            """);
+
+        static void verify(ModuleSymbol m)
+        {
+            var p = m.GlobalNamespace.GetMember<MethodSymbol>("C.M").Parameters.Single();
+            Assert.Equal(RefKind.RefReadOnlyParameter, p.RefKind);
+            VerifyRequiresLocationAttributeSynthesized(m);
+        }
+    }
+
+    [Fact]
+    public void ManuallyDefinedAttribute()
+    {
+        var source = $$"""
+            class C
+            {
+                public void M(ref readonly int p) { }
+            }
+
+            namespace {{RequiresLocationAttributeNamespace}}
+            {
+                class {{RequiresLocationAttributeName}}
```

Comment:
nit: we probably want to derive from System.Attribute, although it looks like it wasn't technically required